### ISSUE TITLE
Composite types

### ISSUE THREAD

**Issue by kdex** [vitaly-t/pg-promise#162]
I have a very simple composite type named `poi` (Point of Interest):

```sql
CREATE TYPE poi AS (
	"id" BIGINT,
	"element" element
);
```

for which I have defined a constructor as

```sql
CREATE OR REPLACE FUNCTION poi (
	BIGINT,
	element
) RETURNS poi AS $$
	SELECT CAST(ROW($1, $2) AS poi);
$$ LANGUAGE 'sql' IMMUTABLE STRICT;
```

I'd like to use `pg-promise` to insert an array of `poi`s into a database, for which I'd need to call the constructor for each of them behind the scenes.

A resulting query will look something like:

```js
INSERT INTO "some-table" (location, pois) VALUES (ST_MakePoint(50, 60), ARRAY[poi(2442522, 'node'), poi(359582, 'way')]::poi[]);
```

I fail to see how `pgpDB.helpers.insert` could construct such a query, and using template literals comes with the disadvantage that the query's syntax becomes invalid on the first `'`.

Neither do I need this to be a transaction, nor do I want to split this up into multiple single `INSERT`s. Can `pg-promise` do that?

**Comment by vitaly-t** [comment]
Do I understand it right that you want to execute just one insert with two columns?

**Comment by kdex** [comment]
Yes. There's three columns in the original source, but I've simplified the scenario for GitHub.

**Comment by vitaly-t** [comment]
The tricky part of your query is the formatting for individual parameters, not for the over query, because you are not doing a multi-row insert, it is just one insert.

Because of that, even though you still can use the `helpers` namespace for the query itself, it is not of much help.

The main thing that you need is the [Custom Type Formatting](https://github.com/vitaly-t/pg-promise#custom-type-formatting) in this case, which can let you format custom objects in any way you want.

You need one type for the first column, another one for the second one...

**Comment by kdex** [comment]
Oh, that looks very promising. I'll give it a try. Will close issue once solved. Thanks!

**Comment by vitaly-t** [comment]
For example, for the first column type you might use:
```js
function STPoint(x, y) {
    this.x = x;
    this.y = y;

    this._rawDBType = true;
    this.formatDBType = function () {
        return pgp.as.format("ST_MakePoint(${x}, ${y})", this);
    };
}
```

**Comment by vitaly-t** [comment]
Accordingly, for the second type you can have this type:
```js
function POI(id, element) {
    this.id = id;
    this.element = element;

    this._rawDBType = true;
    this.formatDBType = function () {
        return pgp.as.format("poin(${id}, ${element})", this);
    };
}
```

**Comment by vitaly-t** [comment]
And here's the complete example:

```js
function STPoint(x, y) {
    this.x = x;
    this.y = y;

    this._rawDBType = true;
    this.formatDBType = function () {
        return pgp.as.format("ST_MakePoint($1, $2)", [this.x, this.y]);
    };
}


function POI(id, element) {
    this.id = id;
    this.element = element;

    this._rawDBType = true;
    this.formatDBType = function () {
        return pgp.as.format("poin($1, $2)", [this.id, this.element]);
    };
}

var loc = new STPoint(50, 60);
var pois = [
    new POI(2442522, 'node'),
    new POI(359582, 'way')
];

db.none("INSERT INTO $1~(location, pois) VALUES($2, $3::poi[])", ['some-table', loc, pois]);
```

It will execute:
```
INSERT INTO "some-table"(location, pois) VALUES(ST_MakePoint(50, 60), array[poin(2442522, 'node'),poin(359582, 'way')]::poi[])
```

**Comment by vitaly-t** [comment]
F.Y.I. Important breaking change in [v6.5.0](https://github.com/vitaly-t/pg-promise/releases/tag/v.6.5.0).

**Comment by vitaly-t** [comment]
This part of code would need to be changed for v6.5.0 of the library:

```js
function STPoint(x, y) {
    this.x = x;
    this.y = y;
    this._rawType = true;
    this.toPostgres = a => pgp.as.format('ST_MakePoint($1, $2)', [a.x, a.y])
}

function POI(id, element) {
    this.id = id;
    this.element = element;
    this._rawType = true;
    this.toPostgres = a => pgp.as.format('poi($1, $2)', [a.id, a.element])
}
```

**Comment by kdex** [comment]
Thank a lot for saving me the trouble, much obliged! 😃

**Comment by vitaly-t** [comment]
Actually, you can make even much simpler:
```js
function STPoint(x, y) {
    this._rawType = true;
    this.toPostgres = () => pgp.as.format('ST_MakePoint($1, $2)', [x, y])
}

function POI(id, element) {
    this._rawType = true;
    this.toPostgres = () => pgp.as.format('poi($1, $2)', [id, element])
}
```

**Comment by kdex** [comment]
Looking at how one would have to set `_rawType` to `true`, I'm actually a little confused: Traditionally, ECMAScript, for the lack of a real private fields, has been using leading underscores to denote that API users aren't supposed to change a field, as it might come with unexpected side effects.

If, at some point, you decide to transform the codebase to a newer ECMAScript version (for which there *are* plans of supporting true, [private fields](https://github.com/tc39/proposal-class-fields#private-fields)), users wouldn't be able to change `_rawType` anymore.

Why is it denoted like a private variable, although based on the advice you're giving, it seems like people are supposed to use it?

**Comment by vitaly-t** [comment]
`_rawType` isn't private, and cannot be, because it is accessed by the query-formatting engine.

The reason it uses an underscore in front is because it is meant to extend existing objects, and the underscore minimizes the chances of a conflict with an already existing property in the object.

**Comment by kdex** [comment]
Feel free to correct me if I'm wrong, but to me, this sounds like you might benefit from creating a prototype that users are supposed to extend. E.g. something akin to:

```js
class POI extends RawType {
  toPostgres() {
      // …
  }
}
```

This would even get rid of the need for a `_rawType` property altogether, as internally, you could rely on `x instanceof RawType` instead.

(Also feel free to transform this example to ES5 if you don't like syntactical classes.)

**Comment by vitaly-t** [comment]
This will not work on existing objects. It will only work on types that you declare in your code.

**Comment by kdex** [comment]
I don't think I'm following you; could you elaborate on that?

**Comment by vitaly-t** [comment]
With the existing solution you can take any object that was created outside of your own code, and turn it into self-formatting.

With what you suggest, you would have to wrap each external type into a local class before you can use Custom Type Formatting, that would be very awkward.

**Comment by kdex** [comment]
Ah, yes, that makes sense. Though wrapping it isn't really that involved:
```js
Object.setPrototypeOf(obj, RawType.prototype);
```

But before I go as far and recommend you mess with prototypes at runtime, which *will* hurt performance in most current engines, I'll stop. 😄 Thanks for the explanation!

**Comment by vitaly-t** [comment]
It is worth noting that syntax for [Custom Type Formatting](https://github.com/vitaly-t/pg-promise#custom-type-formatting) slightly changed in v6.5.0 :wink:

**Comment by vitaly-t** [comment]
Version [7.2.0](https://github.com/vitaly-t/pg-promise/releases/tag/v.7.2.0) changed it a bit.
